ape_hardhat/__init__.py

```diff
@@ -0,0 +1,237 @@
+try:
+    from importlib.metadata import PackageNotFoundError as _PackageNotFoundError  # type: ignore
+    from importlib.metadata import version as _version  # type: ignore
+except ModuleNotFoundError:
+    from importlib_metadata import PackageNotFoundError as _PackageNotFoundError  # type: ignore
+    from importlib_metadata import version as _version  # type: ignore
+
+try:
+    __version__ = _version(__name__)
+except _PackageNotFoundError:
+    # package is not installed
+    __version__ = "<unknown>"
+
+import atexit
+import ctypes
+import platform
+import random
+import signal
+import sys
+import time
+from subprocess import PIPE, Popen, call
+from typing import Any, List, Optional
+
+from ape import plugins
+from ape_http.providers import DEFAULT_SETTINGS, EthereumProvider, NetworkConfig
+
+EPHEMERAL_PORTS_START = 49152
+EPHEMERAL_PORTS_END = 60999
+HARDHAT_CHAIN_ID = 31337
+HARDHAT_CONFIG = """
+// See https://hardhat.org/config/ for config options.
+module.exports = {
+};
+"""
+HARDHAT_START_NETWORK_RETRIES = [0.1, 0.2, 0.3, 0.5, 1.0]  # seconds between network retries
+HARDHAT_START_PROCESS_ATTEMPTS = 3  # number of attempts to start subprocess before giving up
+
+
+def _signal_handler(signum, frame):
+    """Runs on SIGTERM and SIGINT to force ``atexit`` handlers to run."""
+    atexit._run_exitfuncs()
+    sys.exit(0)
+
+
+def _set_death_signal():
+    """Automatically sends SIGTERM to child subprocesses when parent process dies."""
+    if platform.uname().system == "Linux":
+        # from: https://stackoverflow.com/a/43152455/75956
+        # the first argument "1" is PR_SET_PDEATHSIG
+        # the second argument is what signal to send the child
+        libc = ctypes.CDLL("libc.so.6")
+        return libc.prctl(1, signal.SIGTERM)
+
+
+class HardhatSubprocessError(RuntimeError):
+    pass
+
+
+class HardhatNetworkConfig(NetworkConfig):
```

Comment:
It appears that ape's `ConfigManager` cannot handle optional types?

When I do `ape --config` with the hardhat plugin installed, it fails and thinks all the optional types are `None`.
We should probably file a bug in `ape` for this.

Additionally, `List` types appear to not be supported in the ape config API. I am not sure if this is intentional or not: https://github.com/ApeWorX/ape/blob/main/src/ape/api/config.py#L17-L29

So either 1 or 2 bugs in `ape` prevent this config from working when doing `ape --config`

Reply:
Both are bugs with the config setup in `ape`